### PR TITLE
NDRS-1114: Step Information should be present on the event stream

### PR DESCRIPTION
**Pull request by darthsiroftardis**
REF: https://casperlabs.atlassian.net/browse/NDRS-1114

CHANGELOG:

- Added a new `ContractRuntimeAnnouncement` that raises an event when a step has been committed
- Added a new SSE Event that puts the committed `ExecutionEffect` to the event stream
- Additional wire up to generate this event

Sample output attached

[ndrs1114.txt](https://github.com/casper-network/casper-node/files/6379671/ndrs1114.txt)

### CODE REVIEW

File: node/src/components/contract_runtime.rs [hunk 1]
```diff
@@ -336,8 +336,8 @@ where
                             metrics
                                 .commit_upgrade
                                 .observe(start.elapsed().as_secs_f64());
-                            trace!(?result, "upgrade result");
-                            responder.respond(result).await
+                            info!(?result, "upgrade result");
+                            responder.respond(result.clone()).await
```
Comment (xcthulhu) [inline node/src/components/contract_runtime.rs:340]: I don't understand why we have to clone here...

File: node/src/effect.rs [hunk 2]
```diff
@@ -130,6 +130,7 @@ use requests::{
 };
 
 use self::announcements::BlocklistAnnouncement;
+use casper_execution_engine::core::engine_state::execution_effect::ExecutionEffect;
```
Comment (darthsiroftardis) [inline node/src/effect.rs:133]: I note that this out of place, I shall reorder

File: node/src/components/event_stream_server/event.rs [hunk 3]
```diff
@@ -38,6 +42,11 @@ impl Display for Event {
                 public_key, timestamp, era_id,
             ),
             Event::FinalitySignature(fs) => write!(formatter, "finality signature {}", fs),
+            Event::Step { era_id, effect } => write!(
+                formatter,
+                "Execution effect: {:?} committed in era: {}",
```
Comment (Fraser999) [inline node/src/components/event_stream_server/event.rs:47]: `Display` implementations generally shouldn't use `Debug` output internally.  Maybe we should remove this `{:?}` and ticket an update for EE?

Also, `Display` impls should generally be lowercase, and the output doesn't mention "step" anywhere, so I'd change this to something like `"step completed for {}", era_id`.

File: node/src/effect.rs (hunk 2 above)
Comment (Fraser999) [inline node/src/effect.rs]: This import is in the wrong location.

File: node/src/effect.rs [hunk 4]
```diff
@@ -695,6 +696,22 @@ impl<REv> EffectBuilder<REv> {
             .await
     }
 
+    /// Announce a committed Step result
+    pub(crate) async fn announce_step_result(self, era_id: EraId, execution_effect: ExecutionEffect)
```
Comment (Fraser999) [inline node/src/effect.rs:700]: It's not so much a step result as a step success:
```suggestion
    /// Announce a Step success.
    pub(crate) async fn announce_step_success(self, era_id: EraId, execution_effect: ExecutionEffect)
```

File: node/src/effect/announcements.rs [hunk 5]
```diff
@@ -222,6 +222,8 @@ pub enum ContractRuntimeAnnouncement {
     LinearChainBlock(Box<LinearChainBlock>),
     /// A block was requested to be executed, but it had been executed before.
     BlockAlreadyExecuted(Box<Block>),
+    /// A Step result was committed and has altered global state.
+    StepCommitted(EraId, ExecutionEffect),
```
Comment (Fraser999) [inline node/src/effect/announcements.rs:226]:
```suggestion
    /// A Step succeeded and has altered global state.
    StepSuccess {
        era_id: EraId,
        execution_effect: ExecutionEffect,
    },
```

File: node/src/effect/announcements.rs [hunk 6]
```diff
@@ -239,6 +241,11 @@ impl ContractRuntimeAnnouncement {
     pub fn block_already_executed(block: Block) -> Self {
         Self::BlockAlreadyExecuted(Box::new(block))
     }
+
+    /// Create a ContractRuntimeAnnouncement::StepCommitted from an execution effect.
+    pub fn step_result_success(era_id: EraId, execution_effect: ExecutionEffect) -> Self {
```
Comment (Fraser999) [inline node/src/effect/announcements.rs:246]:
```suggestion
    /// Create a ContractRuntimeAnnouncement::StepSuccess from an execution effect.
    pub fn step_success(era_id: EraId, execution_effect: ExecutionEffect) -> Self {
```

File: node/src/effect/announcements.rs [hunk 7]
```diff
@@ -263,6 +270,13 @@ impl Display for ContractRuntimeAnnouncement {
             ContractRuntimeAnnouncement::BlockAlreadyExecuted(block) => {
                 write!(f, "block had been executed before: {}", block.hash())
             }
+            ContractRuntimeAnnouncement::StepCommitted(era_id, ..) => {
+                write!(
+                    f,
+                    "execution effect has been successfully committed in era: {}",
```
Comment (Fraser999) [inline node/src/effect/announcements.rs:276]: Note that the `Display` impl for `EraId` already prints the word "era", so no need to output it again:
```suggestion
                    "step completed for {}",
```

**Review by Fraser999 — COMMENTED** [review]
Just some nitpicks.  Otherwise looking good.

File: node/src/components/contract_runtime.rs [hunk 8]
```diff
@@ -325,7 +325,7 @@ where
                         upgrade_config,
                         responder,
                     } => {
-                        trace!(?upgrade_config, "upgrade");
+                        info!(?upgrade_config, "upgrade");
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/contract_runtime.rs:328]: Yeah @xcthulhu and @dwerner were helping me to see if I could bubble the upgrade step results. But your point is well made, I can bring this back down to trace I suppose

[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/contract_runtime.rs:328]: They suggested it, not requested, sorry for the ambiguity

File: execution_engine/src/core/engine_state/mod.rs [hunk 9]
```diff
@@ -2031,7 +2031,7 @@ where
             .commit(
                 correlation_id,
                 step_request.pre_state_hash,
-                effects.transforms,
+                effects.clone().transforms,
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply execution_engine/src/core/engine_state/mod.rs]: Applied suggestion

File: node/src/components/event_stream_server/event.rs (hunk 3 above)
Comment (darthsiroftardis) [reply node/src/components/event_stream_server/event.rs]: Applied suggestion

File: node/src/effect.rs (hunk 2 above)
Comment (darthsiroftardis) [reply node/src/effect.rs]: Moved

File: node/src/effect.rs (hunk 4 above)
Comment (darthsiroftardis) [reply node/src/effect.rs]: Applied suggestion

File: node/src/effect/announcements.rs (hunk 5 above)
Comment (darthsiroftardis) [reply node/src/effect/announcements.rs]: Applied suggestion

File: node/src/effect/announcements.rs (hunk 7 above)
Comment (darthsiroftardis) [reply node/src/effect/announcements.rs]: Applied suggestion

File: node/src/effect/announcements.rs (hunk 6 above)
Comment (darthsiroftardis) [reply node/src/effect/announcements.rs]: Applied suggestion

File: node/src/components/contract_runtime.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fraser999) [reply node/src/components/contract_runtime.rs:328]: Let's split the difference and go to `debug!` then?  That way, it's not likely to pollute normal users' logs.

File: node/src/components/event_stream_server/event.rs [hunk 10]
```diff
@@ -38,6 +42,7 @@ impl Display for Event {
                 public_key, timestamp, era_id,
             ),
             Event::FinalitySignature(fs) => write!(formatter, "finality signature {}", fs),
+            Event::Step { era_id, .. } => write!(formatter, "step committed for era: {}", era_id),
```
Comment (Fraser999) [inline node/src/components/event_stream_server/event.rs:45]:
```suggestion
            Event::Step { era_id, .. } => write!(formatter, "step committed for {}", era_id),
```

File: node/src/components/contract_runtime.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/contract_runtime.rs:328]: Gotcha!!

File: node/src/components/event_stream_server/event.rs (hunk 10 above)
Comment (darthsiroftardis) [reply node/src/components/event_stream_server/event.rs]: Applied suggestion

**Review by Fraser999 — APPROVED** [review]

**Comment by darthsiroftardis** [comment]
bors merge

**Pull request merged**